redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/ProtocolNode.kt

```diff
@@ -85,8 +91,19 @@ public class ProtocolChildren<W : Any>(
     }
   }
 
-  internal fun remove(index: Int, count: Int) {
+  /**
+   * @param subtreeIds Invoked with the ID of each removed widget and all widgets contained within
+   * its children, recursively.
+   * @param removedNode Invoked with each removed node on this instance (i.e., not recursively).
+   */
+  internal fun remove(index: Int, count: Int, subtreeIds: (Id) -> Unit, removedNode: (ProtocolNode<W>) -> Unit) {
```

Comment:
This function signature is awkward . . . it’s designed to support exactly one callsite, but defined here and not at that callsite. I think it’d be simpler to move the behavior there?

Reply:
(I’m also a little bit anxious about the amount of functions and IDs being boxed to support this faux encapsulation?)

Reply:
Yeah it's very much not great, but the entire protocol-based system is really a single unit that's all aware of each other. This "design" (or non-design, really) is in service of avoiding the need to allocate anything.

The other one I built removed the native views (`children.remove(..)`) and then returned the `nodes.sublist` to the caller for it to consume, but then to also clear, which was also weird ownership-wise. Anything else I built had a bunch of allocation.

In the guest, the relationship is inverted where the nodes hold a shared state. In the host, the state holds the nodes, so there's no way for the nodes to mutate the state directly like the guest does.

Reply:
I didn't realize `nodes` was internal and already used by your reuse stuff. That will allow moving it to the caller and avoiding lambdas altogether.